File: packages/sylo-web-access/extensions/fetch-headless.ts
/**
 * F2 fetch tier — Crawl4AI via `scripts/fetch_headless.py`.
 */
import { asRecord, runPythonScript } from './python-runner.ts'

export interface HeadlessFetchOutcome {
  ok: true
  url: string
  title: string
  markdown: string
  adequate: boolean
  inadequateReason?: string
  tier: string
  screenshotB64?: string
}

export interface HeadlessFetchFailure {
  ok: false
  url: string
  error: string
}

/**
 * Headless render + markdown extraction; optional viewport screenshot (base64).
 */
export async function fetchHeadless(
  url: string,
  options: { screenshot: boolean },
): Promise<HeadlessFetchOutcome | HeadlessFetchFailure> {
  const args = ['--url', url]
  if (options.screenshot) args.push('--screenshot')
  const py = await runPythonScript('fetch_headless.py', args)
  if (!py.ok) return { ok: false, url, error: py.error }
  const data = asRecord(py.data)
  if (!data?.ok) {
    return { ok: false, url, error: String(data?.error ?? 'F2 fetch failed') }
  }
  const markdown = String(data.markdown ?? '').trim()
  return {
    ok: true,
    url: String(data.url ?? url),
    title: String(data.title ?? url),
    markdown,
    adequate: data.adequate === true,
    inadequateReason:
      typeof data.inadequate_reason === 'string' ? data.inadequate_reason : undefined,
    tier: String(data.tier ?? 'F2'),
    screenshotB64:
      typeof data.screenshot_b64 === 'string' && data.screenshot_b64.length > 0 ?
        data.screenshot_b64
      : undefined,
  }
}

File: packages/sylo-web-access/scripts/fetch_headless.py
#!/usr/bin/env python3
"""F2 fetch tier — Crawl4AI headless render + markdown + optional viewport screenshot (stdout JSON)."""
from __future__ import annotations

import argparse
import json
import sys


async def run_fetch(url: str, screenshot: bool) -> dict:
    from crawl4ai import AsyncWebCrawler, CacheMode, CrawlerRunConfig

    config = CrawlerRunConfig(
        cache_mode=CacheMode.BYPASS,
        screenshot=screenshot,
        force_viewport_screenshot=True,
        screenshot_wait_for=1.0,
        page_timeout=45_000,
    )
    async with AsyncWebCrawler() as crawler:
        result = await crawler.arun(url=url, config=config)

    if not result.success:
        return {
            "ok": False,
            "url": url,
            "error": result.error_message or "crawl_failed",
            "escalate": False,
        }

    markdown = (result.markdown or "").strip()
    title = ""
    if getattr(result, "metadata", None) and isinstance(result.metadata, dict):
        title = str(result.metadata.get("title") or "").strip()

    shot = None
    if screenshot and result.screenshot:
        shot = result.screenshot

    adequate = len(markdown) >= 500
    return {
        "ok": True,
        "url": url,
        "title": title or url,
        "markdown": markdown,
        "adequate": adequate,
        "inadequate_reason": None if adequate else "F2 markdown under 500 chars",
        "screenshot_b64": shot,
        "tier": "F2",
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--url", required=True)
    parser.add_argument(
        "--screenshot",
        action="store_true",
        help="Capture viewport PNG as base64",
    )
    args = parser.parse_args()

    try:
        import asyncio
    except ImportError as exc:
        print(json.dumps({"ok": False, "error": str(exc)}))
        return 1

    try:
        payload = asyncio.run(run_fetch(args.url, args.screenshot))
    except ImportError as exc:
        print(
            json.dumps(
                {
                    "ok": False,
                    "error": f"crawl4ai not installed: {exc}",
                    "escalate": False,
                }
            )
        )
        return 1
    except Exception as exc:  # noqa: BLE001
        print(json.dumps({"ok": False, "url": args.url, "error": str(exc), "escalate": False}))
        return 1

    print(json.dumps(payload))
    return 0 if payload.get("ok") else 1


if __name__ == "__main__":
    sys.exit(main())
